### DISCUSSION TITLE
What's the optimal configuration/workflow when runner updates are disabled?

### DISCUSSION THREAD

**Discussion by chrisforrette** [actions/actions-runner-controller#1914] [Questions] [answered]
I've got the actions runner controller configured to operate very ephemerally:

* In response to webhooks with this autoscaler configuration:

```
scaleUpTriggers = [{
      # Scale up an instance when the webhook receives a `workflow_job` event from GitHub
      githubEvent = {
        workflowJob = {}
      }
      # Amount of time an instance will remain idle while waiting for another job before terminating
      duration = "30m0s"
}]
```
* With `ephemeral: true` set

And I've found that sometimes pods will spin up and start the auto update process (with a `Runner update in progress, do not shutdown runner.` log), then the container stops when it's done and doesn't actually seem to process any jobs, leaving the job that triggered it hanging and waiting for another pod.

So I tried setting `DISABLE_RUNNER_UPDATE` to `true` and then I get this in the logs: `An error occurred: Runner version v2.289.1 is deprecated and cannot receive messages.` The `v2.2989.1` appears to be the [GitHub Actions Runner version](https://github.com/actions/runner/releases).

This made me wonder: how do updates to the GitHub Actions Runner actually work? And is there a way to schedule them out of band of processing incoming jobs?

Thanks!

**Comment by toast-gear** [comment, marked as answer]
Whilst GitHub let you disable the self-update process it doesn't mean you can run an old version forever. You have 30 days to upgrade to the latest release upon a new version being released, this is a GitHub limitation https://github.blog/changelog/2022-02-01-github-actions-self-hosted-runners-can-now-disable-automatic-updates/

```
For compatibility with the GitHub Actions service, you will need to manually update your runner within 30 days 
of a new runner version being available.
```

The reason for this, what on the surface seems like an arbitrary limitation, is tonnes of the logic for new features is in the acitons/runner software and not solely server-side in github.com code. As a result they only let you _delay_ the update so you can do it in a controlled manner rather than it being done at random and in the process breaking one of the core philosophies behind containers.

New images from us get produced when Renovate notices there is a new release.

Unfortunately the community hasn't produced a `show and tell` demonstration of how they are automating this process but it wouldn't be super hard to do between renovate/dependabot/a script on a cron, auto-merge and workflow dispatches. It is somewhat environment specific though so we can't really provide a general template. It would be amazing if the community started to build up their implementations as `show and tell` entries in dicussions so we can see how people are solving the problem through automation.

> And I've found that sometimes pods will spin up and start the auto update process (with a Runner update in progress, do not shutdown runner. log), then the container stops when it's done and doesn't actually seem to process any jobs, leaving the job that triggered it hanging and waiting for another pod.

We've had reports of the self-update process resulting on runners unable to pick up jobs for hours before. I'd say having the runners self-update is probably not the way to go and instead automating promoting images with the newer version of the actions/runner software is the direction you should head.

https://github.com/actions-runner-controller/actions-runner-controller/blob/master/TROUBLESHOOTING.md#delay-in-jobs-being-allocated-to-runners

> And is there a way to schedule them out of band of processing incoming jobs?

I was pretty sure there was an issue raised against the actions/runner repository for being able to schedule updates but I can't find one, perhaps it was just a comment on an issue instead, regardless this isn't a feature in the actions/runner software unfortunately.

**Reply by chrisforrette** [reply]
@toast-gear Thank you for the thorough response, this is super helpful!

I realized we've been using the `latest` tag from this very unofficial-looking public ECR repo and it looks like their tags are generally up-to-date but `latest` hasn't been updated in 7 months 😬 https://gallery.ecr.aws/c2u4x1k5/summerwind/actions-runner ...So that explains why the GHA runner keeps getting updated. Y'all don't have an official public ECR repo do you?

Thanks again!

**Reply by toast-gear** [reply]
Nope, we currently only publish to DockerHub and GitHub Packages:

https://hub.docker.com/r/summerwind/actions-runner
https://hub.docker.com/r/summerwind/actions-runner-dind
https://hub.docker.com/r/summerwind/actions-runner-dind-rootless
https://github.com/orgs/actions-runner-controller/packages?repo_name=actions-runner-controller

https://gallery.ecr.aws/c2u4x1k5/summerwind/actions-runner Wowzer! 76K+ downloads from just some random person's mirror!

**Reply by chrisforrette** [reply]
> Wowzer! 76K+ downloads from just some random person's mirror!

I thought the same thing 😄 Thanks again!
